mockwebserver/src/main/kotlin/mockwebserver3/MockWebServer.kt

```diff
@@ -449,33 +446,33 @@ public class MockWebServer : Closeable {
       if (!processTunnelRequests()) return
 
       val protocol: Protocol
-      val socket: Socket
+      val socket: MockWebServerSocket
       when {
         sslSocketFactory != null -> {
           if (firstExchangePeek.failHandshake) {
             dispatchBookkeepingRequest(
               connectionIndex = connectionIndex,
               exchangeIndex = nextExchangeIndex++,
-              socket = raw,
+              socket = MockWebServerSocket(raw),
             )
             processHandshakeFailure(raw)
             return
           }
-          socket =
+          val sslSocket =
             sslSocketFactory!!.createSocket(
               raw,
               raw.inetAddress.hostAddress,
               raw.port,
               true,
-            )
-          val sslSocket = socket as SSLSocket
+            ) as SSLSocket
+          socket = MockWebServerSocket(sslSocket)
           sslSocket.useClientMode = false
           if (clientAuth == CLIENT_AUTH_REQUIRED) {
             sslSocket.needClientAuth = true
           } else if (clientAuth == CLIENT_AUTH_REQUESTED) {
             sslSocket.wantClientAuth = true
           }
-          openClientSockets.add(socket)
+          openClientSockets.add(sslSocket)
```

Comment:
worth putting MockWebServerSocket in openClientSockets?

Reply:
It’s a bit annoying ’cause we don’t create the `MockWebServerSocket` until we know if it’s SSL or not, but we need it in that collection before that.